### DISCUSSION TITLE
How to write network volume driver?

### DISCUSSION THREAD

**Discussion by abitrolly** [containers/podman#16259] [Q&A]
I want to fetch volume contents from the network using custom protocol. Where should I start and what do I need to add that ability to podman?

**Comment by mheon** [comment]
Podman supports the volume plugin protocol described [here](https://docs.docker.com/engine/extend/plugins_volume/). Writing a plugin to this specification should do what you want.

**Reply by abitrolly** [reply]
Not quite what I was looking for.

> it must provide the Docker Daemon with writeable paths on the host filesystem. The Docker daemon provides these paths to containers to consume. The Docker daemon makes the volumes available by bind-mounting the provided paths into the containers.

Does that mean there will be the same problems with SELinux as with :z :Z options? Do I need to care about potential inability to read filesystem if unprivileged container runs with its own USER?

**Reply by rhatdan** [reply]
That would totally be up to the volume driver.  If the file system supports XAttrs then :z and :Z would probably work, as well as :U.

**Reply by abitrolly** [reply]
If I add OS FS layer to the cake, I will have to deal with various SELinux problems, and my users won't be able to write to volumes if their container is unprivileged and uses different USER than their own - https://github.com/containers/podman/discussions/16258

I need another solution that doesn't touch OS FS.
